[23.1] Fix delete collection + elements

davelopez opened the pull request:
Fixes #16866

Before this fix, deleting a collection will only mark the HDCA as deleted but not the contained elements. There are currently 2 ways of deleting collections from the History Panel:

### Using the "Trash" icon in a collection

  We used to display (in the old history) a menu when deleting a single collection to choose to also delete its contents, this menu is now back to help mitigate this issue https://github.com/galaxyproject/galaxy/issues/16866#issuecomment-1768097231
    ![image](https://github.com/galaxyproject/galaxy/assets/46503462/5c2e1155-8360-4e94-8687-610d30b486c6)

### Using the "bulk delete" operation and including collections in the selection

  When deleting collections in bulk, we cannot present this choice for the user, we need to decide the default behavior (delete elements too vs only HCDA). Since using the "bulk purge" when there are collections in the selection will also purge its items, for consistency, using the "bulk delete" will now also mark elements as deleted in the same way.
    ![DeleteCollectionBulk](https://github.com/galaxyproject/galaxy/assets/46503462/a6893ef6-e366-4fdf-b52a-b3ee66d1c000)


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Follow the instructions in #16866

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Display menu when deleting a collection

- Allow to delete only the top collection or also the items contained

davelopez committed:
Delete orphaned and hidden datasets in collection

When deleting the collection via history bulk delete

davelopez committed:
Add API tests for deleting collections in bulk

davelopez commented on lib/galaxy/managers/hdcas.py:
```diff
@@ -103,6 +108,23 @@ def update_attributes(self, content, payload: Dict):
+            model.DatasetCollectionElement.hda_id == dataset.id,
+        )
+        usage_count = self.session().execute(stmt).scalar()
+        return usage_count > 1
```
Is this enough to check that we can mark the dataset as deleted?
I just realized that if we have 2 collections pointing to the same datasets and we delete one collection this will not mark the datasets as deleted (which is what we want) but, if we then delete the other collection, we are still not marking the datasets as deleted. Should we filter out those DCEs whose ancestor HDCA is already deleted for the usage count?

davelopez replied:
In [1ea7ee0](https://github.com/galaxyproject/galaxy/pull/16879/commits/1ea7ee0605261677c81008f26e59e4a15097f195) I have filtered out DCEs that belong to a deleted HDCA from the usage count, so when you delete the last collection they will be marked as deleted too.

davelopez replied:
This no longer applies

davelopez committed:
Exclude from usage DCEs contained in deleted collections

davelopez committed:
Refactor API tests

jdavcs commented on lib/galaxy_test/api/test_history_contents.py:
```diff
@@ -1026,6 +1026,73 @@ def test_purging_collection_should_purge_contents(self):
+
+            # Check datasets are still not deleted
+            history_contents = self._get_history_contents(history_id)
+            for item in history_contents:
```
Given that we reassign `history_contents`, should we first assert that it's not empty and has the expected number of items?

davelopez replied:
Thanks! I added an assert now for this in both cases.

jdavcs replied:
Great! Thank you!

jdavcs commented on lib/galaxy_test/api/test_history_contents.py:
```diff
@@ -1026,6 +1026,73 @@ def test_purging_collection_should_purge_contents(self):
+            self._assert_bulk_success(bulk_operation_result, 1)
+
+            # Check datasets are deleted (and collections too)
+            history_contents = self._get_history_contents(history_id)
```
same as above? So we can be reasonably certain that we are examining the same set of items at each step.

davelopez committed:
Add asserts on history contents in tests

davelopez committed:
Unify bulk delete/purge behavior on collections

Use the same default behavior as "bulk purge" and mark also elements as deleted.

jdavcs reviewed (approved):
I think this is a simple and clean solution which addresses the referenced issue in straightforward way.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1415,9 +1415,14 @@ def _hide(self, item: HistoryItemModel):
-        manager.delete(item, flush=self.flush)
+    def _delete(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        if getattr(item, "deleted", False):
+            # TODO: remove this `update` when we can properly track the operation results to notify the history
```
I don't know if I understand the TODO 😅. We get here if the item is already deleted, so it seems like we wouldn't need to do anything here ?

davelopez replied:
You are absolutely right, too much copy&paste :sweat_smile: this was only necessary for the collection `purge` as there, the purge was not triggering the change in the history update time making the operation seem to be running until something else bumped the history update (if I recall correctly).

Thanks for the catch, I'll remove it.

mvdbeek commented on client/src/components/History/Content/ContentOptions.vue:
```diff
@@ -43,8 +43,21 @@
+                <template v-slot:button-content>
+                    <icon icon="trash" />
+                </template>
+                <b-dropdown-item title="Delete collection Only" @click.prevent.stop="onDeleteItem">
```
```suggestion
                <b-dropdown-item title="Delete collection only" @click.prevent.stop="onDeleteItem">
```

davelopez committed:
Fix typo

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Remove unnecessary update if the element was already deleted in bulk
